### PR TITLE
perf(api): perform EMPs calls in batch

### PR DESCRIPTION
**Pull request by amateima**
Signed-off-by: amateima <amatei@umaproject.org>

**Motivation**

https://app.shortcut.com/uma-project/story/1502/use-multicall-to-prevent-high-infura-usage-for-api-calls-to-contracts

**Summary**

This PR introduces the Multicall2 contract for getting the state of an EMP contract by wrapping the calls into batches instead of using multiple `asyncValues()` calls.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [x]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested

### CODE REVIEW

File: packages/api/src/services/emp-state.ts [hunk 1]
```diff
@@ -1,95 +1,72 @@
...
+    collateralAddresses,
+    syntheticAddresses,
+    multicall2,
+  } = appState;
   const profile = Profile(config.debug);
+  // default props we want to query on contract
+  const staticProps: [string, (x: any) => any][] = [
+    ["priceIdentifier", parseBytes],
+    ["expirationTimestamp", toNumber],
+    ["withdrawalLiveness", toNumber],
```
Comment (amateima) [inline packages/api/src/services/emp-state.ts:35]: @daywiss The current emp state attribute is named `withdrawLiveness`, but now the name is changed to the actual name found in the contract: `withdrawalLiveness `. How should we deal with this? Rename it to the old value or update the emp state type with the new name and inform the clients of the API about this change

Comment (daywiss) [reply packages/api/src/services/emp-state.ts:35]: hey thanks for finding this. i would say change it to align with the contract. its a minor change that does not affect anything in use (not used in umaverse)

File: packages/api/src/apps/api/app.ts [hunk 2]
```diff
@@ -17,6 +17,7 @@ export default async (env: ProcessEnv) => {
   assert(env.EXPRESS_PORT, "requires EXPRESS_PORT");
   assert(env.zrxBaseUrl, "requires zrxBaseUrl");
   assert(env.MULTI_CALL_ADDRESS, "requires MULTI_CALL_ADDRESS");
+  assert(env.MULTI_CALL_2_ADDRESS, "requires MULTI_CALL_2_ADDRESS");
```
Comment (daywiss) [inline packages/api/src/apps/api/app.ts:20]: should we just require a multicall 2 address and use it for both multicall1 and 2? Also i dont remember if multicall 2 is just a drop in replacement for 1, maybe we should just replace the `multicall` reference with the `multicall2` implementation?

File: packages/api/src/services/emp-state.ts (hunk 1 above)
Comment (daywiss) [reply packages/api/src/services/emp-state.ts:35]: if you want we can make a separate issue for this

**Review by daywiss — COMMENTED** [review]
this looks great.  were you able to test this and if so did you see any improvements?

File: packages/api/src/libs/utils.ts [hunk 3]
```diff
@@ -131,6 +131,28 @@ export const BatchRead = (multicall: uma.Multicall) => async (
   );
 };
 
+export const BatchReadWithErrors = (multicall2: uma.Multicall2) => async (
+  calls: [string, (x: any) => any][],
```
Comment (mrice32) [inline packages/api/src/libs/utils.ts:135]: Annoyingly, typescript doesn't have strong enough inference for us to make this a generic type and have typescript infer the return type for the user :/.

**Review by mrice32 — APPROVED** [review]
LGTM! Very nice and clean!

**Comment by chrismaree** [comment]
nit: this PR is not titled in the standard format

File: packages/api/src/apps/api/app.ts (hunk 2 above)
Comment (amateima) [reply packages/api/src/apps/api/app.ts:20]: we can definitely do that

**Comment by amateima** [comment]
> nit: this PR is not titled in the standard format

`perf` is just a broken down type of `improve`. It's specified in the Conventional Commits standard

**Comment by amateima** [comment]
> this looks great. were you able to test this and if so did you see any improvements?

I ran the api locally. The EMP state for all contracts was updated 24s faster (`2m 0s` vs `1m36s`)

File: packages/api/src/libs/utils.ts [hunk 4]
```diff
@@ -131,6 +131,28 @@ export const BatchRead = (multicall: uma.Multicall) => async (
...
+  // multicall batch takes array of {method} objects
+  const results = await multicall2
+    .batch(
+      contract,
+      calls.map(([method]) => ({ method }))
+    )
+    .readWithErrors();
+  // convert results of multicall, an array of responses, into a key value, keyed by contract method
+  return Object.fromEntries(
+    lodash.zip(calls, results).map(([method, result]) => {
```
[the remark this answers is not in the document]
Comment (amateima) [reply packages/api/src/libs/utils.ts]: I changed the naming 👍

**Review by daywiss — APPROVED** [review]

**Pull request merged**